### ISSUE TITLE
Mixed traditional and Asciidoclet javadoc markup

### ISSUE THREAD

**Issue by sebersole** [asciidoctor/asciidoclet#90]
In Hibernate we are trying to migrate to using Asciidoclet to author Javadoc.  However, as Hibernate is a huge project it is simply not feasible to convert all of the docs at once so we have been working on a phased approach change pieces at a time such that we end up with a mix of traditional and Asciidoclet javadoc markup within the same project.  

But in practice this has led to problems rendering the javadocs.  If we enable Asciidoclet it messes up some of the traditional Javadoc markup; e.g. it escapes and "passes through" all HTML tags (you literally end up with `<p/>`, `<li>`, etc in the *rendered* HTML output).  And if we disable Asciidoclet, of course the standard doclet butchers the  Asciidoctor markup.

Are we missing something here?  Does a project have to use one or the other approach?  If so, that is severely limiting for larger projects to move to Asciidoclet

**Comment by sebersole** [comment]
Guillaume had ended up starting a nabble discussion as well - http://discuss.asciidoctor.org/Migrating-to-Asciidoclet-progressively-td6425.html

**Comment by johncarl81** [comment]
Excellent point @sebersole.  I imagine a solution here would be for Asciidoclet to gracefully get out of the way in certain conditions.  We could add a package or class filter parameter to include or exclude certain sources from processing... would that help?

**Comment by johncarl81** [comment]
Put together a quick example here: https://github.com/johncarl81/asciidoclet/tree/include_exclude

This [includes](https://github.com/johncarl81/asciidoclet/blob/include_exclude/pom.xml#L148-L149) the following regex include and exclude options:
```
--include "org.asciidoctor.*"
--exclude "org.asciidoctor.asciidoclet.Stylesheets"
```
This should allow you to convert to Asciidoclet class-by-class or package-by-package as you wish.

**Comment by gsmet** [comment]
I think some sort of marker would probably be easier for our use case. 

But I could see how includes/excludes could be useful too.

**Comment by johncarl81** [comment]
A marker to indicate class-by-class within the javadoc documentation seems noisy especially considering this should be a bridge case for converting to use Asciidoclet.

**Comment by gsmet** [comment]
What I'm wondering is how we would maintain the list of classes to transform with Asciidoctor.

And refactorings come to mind as problematic.

I don't know if it's feasible or not but I think annotating the class Javadoc with @asciidoc or a similar marker would maybe be easier.
We would have to think about it when writing the doc (and annotate the existing classes) but that would survive a refactoring.

That way, you add the marker when you write the doc and you can forget about it.

Just thinking out loud :).

**Comment by johncarl81** [comment]
Good input @gsmet.  How long do you think the conversion would take?

My hope was that using wildcards would help with refactors since you could convert a package and contained packages at a time to avoid an exhaustive list of every class converted, i.e.: `--include org.hibernate.action.*`, as long as a refactor stays within the same package.  If `--include` is specified then everything outside of the `--include` regex would not be processed by Asciidoclet.

**Comment by johncarl81** [comment]
Yeah, absolutely... that was my thought as well - the `@asciidoclet` tag wouldn't show up in the rendered javadoc and I'm open to naming it `@asciidoc`.  Updated [my branch](https://github.com/johncarl81/asciidoclet/tree/include_exclude) with this idea and deployed it to 1.5.7-SNAPSHOT if you want to try it out.

**Comment by gsmet** [comment]
@johncarl81 thanks! I added a couple of comments in your commits.

**Comment by johncarl81** [comment]
https://github.com/asciidoctor/asciidoclet/pull/91

**Comment by gsmet** [comment]
@sebersole could you point me to a class of the 6 code base using Asciidoctor? I'll check how this all works.

thanks!

**Comment by mojavelinux** [comment]
How did the testing go?

I like the `@asciidoc` marker. Is it true that it's only required when a certain flag is set in the configuration? In other words, Asciidoclet defaults to being applied everywhere, but that default can be flipped with an option (hence requiring the `@asciidoc` marker). Is that correct?

**Comment by johncarl81** [comment]
Yes, that's right @mojavelinux - it's only useful when a class is excluded from being processed.  For instance, we can either explicitly include classes for processing or exclude classes from processing, the resulting classes not processed can finally be added back into the set being processed by adding the `@asciidoc` javadoc tag.

**Comment by mojavelinux** [comment]
Thanks for the clarification! That's exactly what I was hoping you'd say.

**Comment by johncarl81** [comment]
@mojavelinux, @msgilligan: it might be early as we haven't gotten feedback, but would you review https://github.com/asciidoctor/asciidoclet/pull/91 and merge?

@sebersole, @gsmet: have you been able to use this feature, and, if so, does it meet your needs?

**Comment by gsmet** [comment]
@johncarl81 sorry about the delay... totally forgot about it.

We plan to release a new version next Wednesday so. I'll check what you did later today or tomorrow. Do you think you could plan a release once you have our feedback?

Thanks!

**Comment by sebersole** [comment]
@mojavelinux, I actually missed the notification from this discussion.  I am working on a benchmark for our current work that is super critical, but I will take a look at this again soon.  But based on the comment by @johncarl81 that sounds perfect

**Comment by mojavelinux** [comment]
Thanks for chiming in @gsmet and @sebersole. Sounds like we're all set to move forward. :tada:

**Comment by sebersole** [comment]
Was there a release of this yet to test out?

**Comment by gsmet** [comment]
@sebersole no you need to build a snapshot. I'll share my ORM branch with you so that you can see what I did

**Comment by gsmet** [comment]
@sebersole see https://github.com/sebersole/hibernate-core/pull/158 . I added instructions in the description.

**Comment by sebersole** [comment]
Is this available to try in any published (SNAPSHOT even) artifact?

**Comment by johncarl81** [comment]
Not on maven central.  You can, of course, build the branch and work with the SNAPSHOT locally.  I'm waiting on is confirmation it fits the use case and review here: https://github.com/asciidoctor/asciidoclet/pull/91

**Comment by sebersole** [comment]
@johncarl81 I hate using mavenLocal.  I'll grab this branch, modify the groupId and publish a SNAPSHOT to the JBoss Snapshot repo and test from there

**Comment by johncarl81** [comment]
If I remember a bit later, I could publish the snapshot to maven central if that helps.

**Comment by sebersole** [comment]
Sorry for the late reply.. I've finally been able to test this out and it does not work for me.  Probably I'm missing something simple.  I use the code from  #91 and build the javadoc - any docs using `@asciidoc` do not render anything.  I saw something above about still needing to specify some configuration property to trigger `@asciidoc` handling - if that's still true, what setting do I need?

**Comment by sebersole** [comment]
What I have atm:

```
configurations {
	asciidoclet {
		description = 'Dependencies for Asciidoctor Javadoc taglet'
	}
}

dependencies {
	// snapshot built from #91
	asciidoclet( libraries.asciidoclet )
}

javadoc {
	configure( options ) {
		docletpath = configurations.asciidoclet.files.asType(List)
		doclet = 'org.asciidoctor.Asciidoclet'
		...
	}

	options.addStringOption( '-exclude-asciidoclet-process', '**' )
}
```

with

```

/**
 * @asciidoclet
 *
 * Represents a selection at the SQL/JDBC level.  Essentially made up of:
 *
 * 		{@link #getJdbcValueExtractor}:: How to read a value from JDBC (conceptually similar to a method reference)
 * 		{@link #getValuesArrayPosition}:: The position for this selection in relation to the "JDBC values array" (see {@link RowProcessingState#getJdbcValue})
 * 		{@link #getJdbcResultSetIndex()}:: The position for this selection in relation to the JDBC object (ResultSet, etc)
 *
 * Additional support for allowing a selection to "prepare" itself prior to first use is defined through
 * {@link #prepare}.  This is generally only used for NativeQuery execution.
 */
public interface SqlSelection {
}
```

Nothing gets rendered for the SqlSelection class-level docs

**Comment by sebersole** [comment]
Could the problem be the use of `org.asciidoctor:asciidoctor-gradle-plugin:1.5.7` in combination with `org.asciidoctor:asciidoclet:1.5.7-SNAPSHOT`?

**Comment by sebersole** [comment]
> Could the problem be the use of `org.asciidoctor:asciidoctor-gradle-plugin:1.5.7` in combination with `org.asciidoctor:asciidoclet:1.5.7-SNAPSHOT`? The later bring a build from #91

Nope that made no difference.  So at this point, as far as I can tell, #91 does not fix our problem - https://github.com/hibernate/hibernate-orm/commit/e43c5a3166c6bd7a1f4dc4b4a617679118846718

**Comment by sebersole** [comment]
Is this implemented?  Can someone please tell me how to use the doclet with mixed Javadoc?

**Comment by sebersole** [comment]
Will the PR ever get integrated?

**Comment by mojavelinux** [comment]
If there are no objections in the next 24-48 hours, I will merge the outstanding PR.

**Comment by johncarl81** [comment]
Sorry I've been absent... Lets go ahead and merge.

**Comment by mojavelinux** [comment]
@johncarl81 Did you have any thoughts about my comment here? https://github.com/asciidoctor/asciidoclet/pull/91#issuecomment-948305206

**Comment by mojavelinux** [comment]
The PR that addresses this issue has now been merged. Thanks @johncarl81!! If there are further enhancements needed, let's discuss them in a separate issue.

**Comment by sebersole** [comment]
Any idea what version will include this?

**Comment by sebersole** [comment]
Is this in any released version yet?  I could not find how to link up issues/prs to releases here.
